File: conway/world.cpp
#include "world.h"
#include <algorithm>
#include <iomanip>
#include <ranges>

using namespace conway;
using std::make_unique;
using std::unordered_map;
using std::ranges::views::iota;

World::World(const int32_t size, const list<Coord> &aliveCells) : size(size),
                                                                  cells(make_unique<unordered_map<Coord, Cell>>()) {
    for (int32_t a : iota(0, size)) {
        for (int32_t b : iota(0, size)) {
            Coord coord = {a, b};
            if (std::find(aliveCells.begin(), aliveCells.end(), coord) != aliveCells.end()) {
                cells->insert(std::make_pair(coord, Cell::Alive));
            } else {
                cells->insert(std::make_pair(coord, Cell::Dead));
            }
        }
    }
}

std::unique_ptr<World> World::evolve() const {
    list<Coord> alive;
    for (auto const &[coord, cell] : *cells) {
        int32_t count = 0;

        for (int32_t a : {-1, 0, 1}) {
            for (int32_t b : {-1, 0, 1}) {
                const Coord currentCoord = {coord.x + a, coord.y + b};
                if (currentCoord != coord &&
                    cells->find(currentCoord) != cells->end() && // https://stackoverflow.com/a/3886599/133764
                    cells->at(currentCoord) == Cell::Alive) {
                    count += 1;
                }
            }
        }

        switch (cell) {
            case Cell::Alive:
                if (count == 2 || count == 3) {
                    alive.push_back(coord);
                }
                break;

            case Cell::Dead:
                if (count == 3) {
                    alive.push_back(coord);
                }
                break;
        }
    }
    return make_unique<World>(size, alive);
}

void World::dump(ostream &os) const {
    os << "\n";

    for (int32_t a : iota(0, size)) {
        if (a == 0) {
            // First line with coordinates
            os << "    ";
            for (int32_t b : iota(0, size)) {
                os << std::setw(3) << b << "|";
            }
            os << "\n";
        }
        os << std::setw(3) << a << "|";
        for (int32_t b : iota(0, size)) {
            os << cells->at({b, a});
        }
        os << "\n";
    }
}

bool World::equals(const World &rhs) const {
    return this->size == rhs.size && std::equal(this->cells->begin(), this->cells->end(), rhs.cells->begin());
}

list<Coord> World::blinker(const Coord &coord) {
    return {
        {coord.x, coord.y},
        {coord.x + 1, coord.y},
        {coord.x + 2, coord.y}
    };
}

list<Coord> World::beacon(const Coord &coord) {
    return {
        {coord.x, coord.y},
        {coord.x + 1, coord.y},
        {coord.x, coord.y + 1},
        {coord.x + 1, coord.y + 1},
        {coord.x + 2, coord.y + 2},
        {coord.x + 3, coord.y + 2},
        {coord.x + 2, coord.y + 3},
        {coord.x + 3, coord.y + 3}
    };
}

list<Coord> World::glider(const Coord &coord) {
    return {
        {coord.x + 2, coord.y + 2},
        {coord.x + 1, coord.y + 2},
        {coord.x, coord.y + 2},
        {coord.x + 2, coord.y + 1},
        {coord.x + 1, coord.y},
    };
}

list<Coord> World::block(const Coord &coord) {
    return {
        {coord.x, coord.y},
        {coord.x + 1, coord.y},
        {coord.x, coord.y + 1},
        {coord.x + 1, coord.y + 1}
    };
}

list<Coord> World::tub(const Coord &coord) {
    return {
        {coord.x + 1, coord.y},
        {coord.x, coord.y + 1},
        {coord.x + 2, coord.y + 1},
        {coord.x + 1, coord.y + 2}
    };
}

File: ConwayLib/World.cs
using System.Text;

namespace ConwayLib;

public class World
{
    public readonly Dictionary<Coord, Cell> Cells = new();
    public int Size { get; }

    public World(int size, ICollection<Coord> aliveCells)
    {
        Size = size;
        for (var a = 0; a < size; ++a)
            for (var b = 0; b < size; ++b)
            {
                var coord = new Coord { X = a, Y = b };
                Cells[coord] = (aliveCells.Contains(coord)) ? Cell.Alive : Cell.Dead;
            }
    }

    public World Evolve()
    {
        var aliveCells = new List<Coord>();
        var range = new[] { -1, 0, 1 };
        foreach (var item in Cells)
        {
            var count = (from a in range
                         from b in range
                         select new Coord { X = item.Key.X + a, Y = item.Key.Y + b })
                .Count(currentCoord => currentCoord != item.Key
                                        && Cells.ContainsKey(currentCoord)
                                        && Cells[currentCoord] == Cell.Alive);

            switch (item.Value)
            {
                case Cell.Alive:
                    if (count == 2 || count == 3) aliveCells.Add(item.Key);
                    break;

                case Cell.Dead:
                    if (count == 3) aliveCells.Add(item.Key);
                    break;
            }
        }

        return new World(Size, aliveCells);
    }

    public override string ToString()
    {
        var builder = new StringBuilder("\n");
        for (var a = 0; a < Size; ++a)
        {
            if (a == 0)
            {
                // First line with coordinates
                builder.Append("    ");
                for (var b = 0; b < Size; ++b) builder.AppendFormat("{0,3}|", b);
                builder.Append('\n');
            }

            builder.AppendFormat("{0,3}|", a);
            for (var b = 0; b < Size; ++b)
            {
                var coord = new Coord { X = b, Y = a };
                var cell = Cells[coord];
                builder.Append(cell.ToEnumString());
            }

            builder.Append('\n');
        }

        return builder.ToString();
    }

    public override bool Equals(object? obj) => obj != null && Equals((World)obj);

    private bool Equals(World other) => Size == other.Size && Cells.SequenceEqual(other.Cells);

    public override int GetHashCode() => HashCode.Combine(Size, Cells);

    public static List<Coord> Blinker(Coord coord) => new()
    {
        new() {X = coord.X,     Y = coord.Y},
        new() {X = coord.X + 1, Y = coord.Y},
        new() {X = coord.X + 2, Y = coord.Y}
    };

    public static List<Coord> Beacon(Coord coord) => new()
    {
        new() {X = coord.X,     Y = coord.Y},
        new() {X = coord.X + 1, Y = coord.Y},
        new() {X = coord.X,     Y = coord.Y + 1},
        new() {X = coord.X + 1, Y = coord.Y + 1},
        new() {X = coord.X + 2, Y = coord.Y + 2},
        new() {X = coord.X + 3, Y = coord.Y + 2},
        new() {X = coord.X + 2, Y = coord.Y + 3},
        new() {X = coord.X + 3, Y = coord.Y + 3}
    };

    public static List<Coord> Glider(Coord coord) => new()
    {
        new() {X = coord.X + 2, Y = coord.Y + 2},
        new() {X = coord.X + 1, Y = coord.Y + 2},
        new() {X = coord.X,     Y = coord.Y + 2},
        new() {X = coord.X + 2, Y = coord.Y + 1},
        new() {X = coord.X + 1, Y = coord.Y}
    };

    public static List<Coord> Block(Coord coord) => new()
    {
        new() {X = coord.X,     Y = coord.Y},
        new() {X = coord.X + 1, Y = coord.Y},
        new() {X = coord.X,     Y = coord.Y + 1},
        new() {X = coord.X + 1, Y = coord.Y + 1}
    };

    public static List<Coord> Tub(Coord coord) => new()
    {
        new() {X = coord.X + 1, Y = coord.Y},
        new() {X = coord.X,     Y = coord.Y + 1},
        new() {X = coord.X + 2, Y = coord.Y + 1},
        new() {X = coord.X + 1, Y = coord.Y + 2}
    };
}
